LoopKitUI/View Controllers/BasalScheduleTableViewController.swift

```diff
@@ -0,0 +1,443 @@
+    }
+
+    private func updateInsertButton() {
+        guard let lastItem = scheduleItems.last else {
+            return
+        }
+        insertButtonItem.isEnabled = scheduleItems.endIndex < maximumScheduleItemCount && !isEditing && lastItem.startTime < lastValidStartTime
+    }
+
+    override func addScheduleItem(_ sender: Any?) {
+        guard !isReadOnly && !isSyncInProgress else {
+            return
+        }
+
+        tableView.endEditing(false)
+
+        var startTime = TimeInterval(0)
+        var value = allowedBasalRates.count > 0 ? allowedBasalRates[0] : 0
+
+        if let lastItem = scheduleItems.last {
+
+            startTime = lastItem.startTime + minimumTimeInterval
+            value = lastItem.value
+
+            if startTime > lastValidStartTime {
+                return
+            }
+        }
+
+        scheduleItems.append(
+            RepeatingScheduleValue(
+                startTime: startTime,
+                value: value
+            )
+        )
+        isScheduleModified = true
+        updateTimeLimitsForItemsAdjacent(to: scheduleItems.endIndex-1)
+
+        super.addScheduleItem(sender)
+
+        updateSyncButton()
+        updateEditButton()
+    }
+
+    override func insertableIndiciesByRemovingRow(_ row: Int, withInterval timeInterval: TimeInterval) -> [Bool] {
+        return insertableIndices(for: scheduleItems, removing: row, with: timeInterval)
+    }
+
+    open override func setEditing(_ editing: Bool, animated: Bool) {
+        tableView.beginUpdates()
+        hideBasalScheduleEntryCells()
+        tableView.endUpdates()
+
+        super.setEditing(editing, animated: animated)
+        updateInsertButton()
+        updateSyncButton()
+    }
+
+    var preferredValueFractionDigits: Int = 1
+
+    public weak var syncSource: BasalScheduleTableViewControllerSyncSource? {
+        didSet {
+            isReadOnly = syncSource?.basalScheduleTableViewControllerIsReadOnly(self) ?? false
+
+            if isViewLoaded {
+                tableView.reloadData()
+            }
+        }
+    }
+
+    private var isSyncInProgress = false {
+        didSet {
+            for cell in tableView.visibleCells {
+                switch cell {
+                case let cell as TextButtonTableViewCell:
+                    cell.isEnabled = !isSyncInProgress
+                    cell.isLoading = isSyncInProgress
+                case let cell as BasalScheduleEntryTableViewCell:
+                    cell.isReadOnly = isCellEditingPossible
+                default:
+                    break
+                }
+            }
+
+            for item in navigationItem.rightBarButtonItems ?? [] {
+                item.isEnabled = !isSyncInProgress
+            }
+
+            navigationItem.hidesBackButton = isSyncInProgress
+        }
+    }
+
+    public var isScheduleValid: Bool {
+        return scheduleItems.endIndex > 0 &&
+            scheduleItems.endIndex <= maximumScheduleItemCount &&
```

Comment:
Nit: *This* one I think is clearer as `scheduleItems.count`, since it's being used to refer to the number of items rather than an index.